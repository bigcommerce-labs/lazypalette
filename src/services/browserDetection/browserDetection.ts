import { isBelowInternetExplorer11 } from './checks/internetExplorer/internetExplorer';
import { UnsupportedBrowser } from './errors/unsupportedBrowser';

export enum BrowserVersion {
    InternetExplorer10 = 'InternetExplorer10',
}

export interface BrowserCheck {
    detected: boolean;
}

export interface BrowserCheckResult {
    browser: BrowserVersion;
    detected: boolean;
}

export interface BrowserDetectionOptions {
    unsupportedBrowsers: BrowserVersion[];
}

const BROWSER_CHECKS = {
    [BrowserVersion.InternetExplorer10]: isBelowInternetExplorer11,
};

const browserDetection = (options: BrowserDetectionOptions) => {
    const { unsupportedBrowsers } = options;

    const browserDetectionChecks = unsupportedBrowsers.map((unsupportedBrowser: BrowserVersion) => {
        return new Promise((resolve, reject) => {
            const check: BrowserCheck = BROWSER_CHECKS[unsupportedBrowser]();

            // Detection works by marking the promise as a failure.
            if (check.detected) {
                const error = `UnsupportedBrowser detected. Detected user is running ${unsupportedBrowser}`;

                return reject(new UnsupportedBrowser(error));
            }

            return resolve({ ...check, browser: unsupportedBrowser });
        });
    });

    // Promise.all returns once all promises are complete or the first promise fails
    return Promise.all(browserDetectionChecks);
};

export default browserDetection;
