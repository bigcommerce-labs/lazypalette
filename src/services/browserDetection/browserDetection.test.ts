import browserDetection, { BrowserDetectionOptions, BrowserVersion } from './browserDetection';

declare let global: any;

describe('browserDetection', () => {
    describe('when we detect the user is running an unsupported browser', () => {
        it('returns a rejected promise with an unsupported browser error', () => {
            Object.defineProperty(global.navigator, 'userAgent', {
                value: 'MSIE 10.0',
                writable: true,
            });

            const options: BrowserDetectionOptions = {
                unsupportedBrowsers: [BrowserVersion.InternetExplorer10],
            };

            expect(browserDetection(options)).rejects.toThrowErrorMatchingSnapshot();
        });
    });

    describe('when we we have many checks which are detected', () => {
        it('returns a rejected promise with an unsupported browser error', () => {
            Object.defineProperty(global.navigator, 'userAgent', {
                value: 'MSIE 10.0; MSIE 9.0;',
                writable: true,
            });

            const options: BrowserDetectionOptions = {
                unsupportedBrowsers: [
                    BrowserVersion.InternetExplorer10,
                    BrowserVersion.InternetExplorer10,
                ],
            };

            expect(browserDetection(options)).rejects.toThrowErrorMatchingSnapshot();
        });
    });

    describe('when we have no unsupported browsers', () => {
        it('returns a list of successful browser checks', () => {
            Object.defineProperty(global.navigator, 'userAgent', {
                value: 'MSIE 10.0',
                writable: true,
            });

            const options: BrowserDetectionOptions = {
                unsupportedBrowsers: [],
            };

            expect(browserDetection(options)).resolves.toEqual([]);
        });
    });

    describe('when we detect the user is running a supported browser', () => {
        it('returns a list of successful browser checks', () => {
            Object.defineProperty(global.navigator, 'userAgent', {
                value: 'Google Chrome :)',
                writable: true,
            });

            const options: BrowserDetectionOptions = {
                unsupportedBrowsers: [BrowserVersion.InternetExplorer10],
            };

            expect(browserDetection(options)).resolves.toEqual([{
                browser: BrowserVersion.InternetExplorer10,
                detected: false,
            }]);
        });
    });

    describe('when we detect the user is running a supported browser and we have many checks', () => {
        it('returns a list of successful browser checks', () => {
            Object.defineProperty(global.navigator, 'userAgent', {
                value: 'Google Chrome :)',
                writable: true,
            });

            const options: BrowserDetectionOptions = {
                unsupportedBrowsers: [
                    BrowserVersion.InternetExplorer10,
                    BrowserVersion.InternetExplorer10,
                    BrowserVersion.InternetExplorer10,
                ],
            };

            expect(browserDetection(options)).resolves.toEqual([
                {
                    browser: BrowserVersion.InternetExplorer10,
                    detected: false,
                },
                {
                    browser: BrowserVersion.InternetExplorer10,
                    detected: false,
                },
                {
                    browser: BrowserVersion.InternetExplorer10,
                    detected: false,
                },
            ]);
        });
    });
});
