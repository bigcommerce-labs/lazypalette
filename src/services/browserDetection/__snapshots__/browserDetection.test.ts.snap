// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browserDetection when we detect the user is running an unsupported browser returns a rejected promise with an unsupported browser error 1`] = `"UnsupportedBrowser detected. Detected user is running InternetExplorer10"`;

exports[`browserDetection when we we have many checks which are detected returns a rejected promise with an unsupported browser error 1`] = `"UnsupportedBrowser detected. Detected user is running InternetExplorer10"`;
